rfcs/0058-secrets-for-services.md

```diff
@@ -0,0 +1,273 @@
+RemainAfterExit=true
+Type=oneshot
+```
+
+Target service:
+
+```
+[Unit]
+Description=Simple test service using a secret
+JoinsNamespaceOf=foo-secrets.service
+
+[Service]
+Environment="[...]"
+
+DynamicUser=true
+ExecStart=/nix/store/3kqc2wmvf1jkqb2jmcm7rvd9lf4345ra-coreutils-8.31/bin/cat /tmp/secret1
+PrivateTmp=true
+```
+
+## NixOS modules integration
+
+To implement an interface as outlined above, a little bit of supporting functionality
+needs to be added somewhere in the nixos library functionality.
+
+An example of some needed functions, of which some could be user exposed configuration,
+is shown in https://github.com/d-goldin/nix-svc-secrets/blob/master/secretslib.nix.
+
+This is mostly functionality containing a _registry_ of existing fetchers, which
+might need to be configured by the user via their system configuration, the
+fetcher logic itself and functionality to generate side-car services and
+expose the secrets scope.
+
+## Rotating secrets
+
+Right now, secrets rotation is not done automatically. When new secrets are
+pushed, it is the responsibility of the user to restart the services affected.
+
+It is assumed that once secrets are rotated, old secrets will become invalid and
+no further harm is done aside from failing to access the resources (and possibly
+restart on its own).
+
+It would be possible to allow for automatic restarts using systemd path monitors.
+Also see _Future work_.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+I can't really think of a serious drawback right now, but hopefully the
+RFC process can surface some.
+
+One aspect is of course the additional number of services generated, but this
+does not seem to be a big issue when using NixOps.
+
+# Alternatives
+[alternatives]: #alternatives
+
+* One approach that has been proposed in the past is a non-world readable store,
+  in issue #8 (support private files in the nix store, from 2012). While this would
+  be pretty great, it's rather complex and has not made progress in a while.
+
+* "Classical" approach of just storing secrets readable only to a service user and
+  utilizing string-paths to reference them. This does not work well anymore with
+  DynamicUser.
+
+* A somewhat similar approach exists in
+  https://cgit.krebsco.de/stockholm/tree/krebs/3modules/secret.nix
+  (loosely related to krops).
+
+* NixOps implements a similar approach, providing a service to expose secrets
+  via a systemd service after it has taken care of deployment.
+
+Impact of not doing this:
+
+Continued proliferation of various, individual solutions, per module and
+depending on the users environment.
+
+Persistent confusion by new-comers and veterans alike, about what the a
+recommended way looks like and a variety of different approaches within
+nixos service modules.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* Is it sufficient to put responsibility on restarting services after key changes
+  onto the user or would an automated mechanism be better?
```

Comment:
I think typically an automated mechanism would be better. Most services just restart gracefully, so an additional unexpected restart is better than a confusing error message and downtime.

Reply:
This very much depends on the service. If I'm dealing with a very heavy service, like for instance an ElasticSearch node, where a restart can cause re-replications or similar it can be very useful to tightly coordinate restarts in a larger setup. So I'm sympathetic to the average case where automatic restart is nice and stress-less but I think if we do it, we should make it configurable to.